### PR TITLE
Test for requiring same strides in adamax.

### PR DESCRIPTION
**Pull request by gchanan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45469 Test for requiring same strides in adamax.**

Adamax used to call max(input, 0, keepdim=False, out=(input, input.new().long()).
But the implementation of the out variant requires that the two tensors have the same strides.

Now we call amax, which fixes the problem, but this was clearly a gap in testing.

Differential Revision: [D23977274](https://our.internmc.facebook.com/intern/diff/D23977274)

### CODE REVIEW

File: test/test_optim.py [hunk 1]
```diff
@@ -1892,5 +1893,29 @@ def test_bn_update_eval_momentum(self):
...
+                super(ModuleWithNonContigWeights, self).__init__()
+                self.weight = torch.nn.Parameter(torch.randn(shape, device=device)
+                                                 .to(memory_format=torch.channels_last))
+
+            def forward(self, input: torch.Tensor) -> torch.Tensor:
+                return input * self.weight
+
+        m = ModuleWithNonContigWeights()
+        input = torch.randn(shape, device=device, requires_grad=True)
+        m(input).sum().backward()
```
Comment (ngimel) [inline test/test_optim.py:1914]: that's a lot of faith in shape propagation working as expected. Can you assert that gradients are actually discontiguous?

**Review by ngimel — APPROVED** [review]

File: test/test_optim.py (hunk 1 above)
Comment (gchanan) [reply test/test_optim.py:1914]: I can, it's not totally clear how the test is supposed to be fixed if that assert fails though.

Comment (gchanan) [reply test/test_optim.py:1914]: I guess I can pre-create the discontiguous inputs, which should reduce the chance of that failing.

Comment (gchanan) [reply test/test_optim.py:1914]: or even better, I can just create the grad tensor without even invoking autograd.
